plugin/plaintextmodel.cpp

```diff
@@ -264,6 +277,46 @@ bool PlainTextModel::readLines(
     return false;
 }
 
+QTextCodec *PlainTextModel::detectCodec(const char *encodedCharacters, qint64 length)
+{
+    QLocale locale;
+    QTextCodec *codec = QTextCodec::codecForLocale();
+
+    UErrorCode error = U_ZERO_ERROR;
+    if (UCharsetDetector * const detector = ucsdet_open(&error)) {
+        ucsdet_setText(detector, encodedCharacters, length, &error);
+
+        int32_t count = 0;
+        if (U_FAILURE(error)) {
+            qWarning() << "Unable to detect text encoding" << u_errorName(error);
+        } else if (const UCharsetMatch ** const matches = ucsdet_detectAll(detector, &count, &error)) {
+            QTextCodec *bestMatch = nullptr;
+            QString localeName = locale.name();
+
+            for (int32_t i = 0; i < count; ++i) {
+                if (QString(ucsdet_getLanguage(matches[i], &error)) == localeName.left(localeName.indexOf("_"))) {
+                    bestMatch = QTextCodec::codecForName(ucsdet_getName(matches[i], &error));
+                    break;
+                }
+            }
+
+            if (bestMatch) {
+                codec = bestMatch;
```

Comment:
Now with utf-8 confidence 100 and language specific locale with confidence 1 will pick the latter, which doesn't feel like the correct thing to do.

Reply:
Well it seems that previous variant was better because because it made a choice according to the codec and to the language. So what was wrong?

Reply:
Corrected

Reply:
Ok, now it handles utf-8 and language specific whichever just has better confidence.

But then again latin1 confidence 100, utf8 or language specific confidence 1 will pick the latter one. Quite heavy bias?

Reply:
Actually the language has better priority because the language of the interface is the native user's language or the language the user can speak anyway. So IMHO language should have better priority then the codec.

Reply:
So I don't understand you. What do you mean?

Reply:
IMHO it would be better to compare with the language (and give the best codec of the locale language)or give just the most confident codec otherwise, because, as I told, the most confident codec could be wrong, so it would be better to mention language

Reply:
First utf8 vs latin1. The latter will be picked up only if detection says there's no chance the content is utf8. If if thinks there's some chance, if only that 1 confidence level versus 100, it will pick now utf8. I don't think there should be any bias between these two, whatever has more confidence should be just chosen. 

And why utf8 is because QTextCodec::codecForLocale() is in practice that for everything these days. In a way would be better not to use the whole method, but instead explicitly create utf8 codec to avoid misunderstandings.

Then the language, this will now in similar way pick encoding that matches the locale, now matter how low confidence there is for it if better guesses don't return the current language. Latin-1with 100 confidence level will still lose against codec with confidence of 1 but matching the language and locale.

Testing this with some files, the language is a bit random: fr, no, ro, tr, he. But the api docs do state that the language might end up guessed as side-effect.

Reply:
Done it

Reply:
Seeing this made the utf-8 now explicit, but all the other problems still remain. Or actually this now seems maybe even worse, e.g. never picks up latin1 no matter how high confidence there is for it? (At least unless the halfway random language guess happens to match current locale.)

Reply:
So if current language is English (e.g) Latin1 with confidence=100 can be detected according to the current language match

Reply:
Sorry, missed these earlier comments, maybe best responding these ones:

> What do you mean? Just language based detection? And why should we multiply confidence which is already calculated and sorted?
> Why cannot I take the best encoding with the language of locale?
> And what should I do if there is no proper language? Perhaps I should take the most confident encoding, shouldn't I? Or the encoding of locale?

So suppose we have latin1 file and the detection returns codec/confidence pairs <latin1, 100>, [...something random], <windows-1251, 3>, <utf8, 1>. What we really want is that latin1, but the code here would return either windows-1251 or utf8 based on the language, both which have extremely low confidence.

However, if the list had <latin1, 100>, <windows-1251, 90> and the current language was russian, a little bias towards the latter could be justified. That's what I meant by multiplying. (Alternatively could be adding something too, but then we'd assume the returned range which didn't seem documented(?)). 90 * 1.5=>135 would lean towards 1251 on this case, but the earlier low confidence 3 * 1.5=> 4.5 would still be low confidence.

So all in all, I'm ok with just the simplistic solution of always just using the first codec returned, but if there's need to have bias towards current language, it should be a reasonable bias and not a forceful override.

Btw, the script of the current locale could probably be more correct way to bias, i.e. giving all the cyrillic languages boost for cyrillic codecs, but that might have a bit more complexity.

Reply:
Do you mean that it would be better to analyse codecs with confidence>100% after multiplying?

Reply:
I'm not talking about any threshold or such. I'm not even sure are these percentages or just plain numbers.  I'm just talking about giving some results a confidence boost if the language matches, and after that picking the codec that has the highest confidence.

Reply:
So what is the purpose of this multiplication? The sequence of codecs won't change and I can find the best language-matched codec with the highest confidence without this multiplication.

Reply:
> However, if the list had <latin1, 100>, <windows-1251, 90> and the current language was russian, a little bias towards the latter could be justified. That's what I meant by multiplying. (Alternatively could be adding something too, but then we'd assume the returned range which didn't seem documented(?)). 90 * 1.5=>135 would lean towards 1251 on this case, but the earlier low confidence 3 * 1.5=> 4.5 would still be low confidence.
> 
Yes, Windows-1251 will go now with 135 confidence, but latin1 will go with 150 confidence. As you see, the sequence of encodings will not change, so what's the purpose? I can find the most confident language-matched encoding without this multiplication

Reply:
Purpose is giving a confidence boost. Making that 90 cyrillic win the 100 latin1. You can find the highest matching already, but you won't have a bias towards anything on it.

Or if you're happy with just the highest confidence, just pick that and ditch all this language matching.

Reply:
> Yes, Windows-1251 will go now with 135 confidence, but latin1 will go with 150 confidence.

So that idea was to give the boost _only_ if the language matches to current locale. If this detects latin1 file containing Russian text, surely it's then better. (Which shouldn't be even possible)

Reply:
Ok. That seems I understood you

Reply:
Done